### PR TITLE
SDK - Containers - Added support for container image cache

### PR DESCRIPTION
**Pull request by Ark-kun**
This change makes `build_image_from_working_dir`  fast when the working directory has not changed between invocations.
We cache pushed container images using specially-calculated context directory hash as the cache key.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/2216)

### CODE REVIEW

File: sdk/python/kfp/containers/_build_image_api.py [hunk 1]
```diff
@@ -111,11 +111,22 @@ def build_image_from_working_dir(image_name: str = None, working_dir: str = None
             dockerfile_text = _generate_dockerfile_text(context_dir, dst_dockerfile_path, base_image)
             with open(dst_dockerfile_path, 'w') as f:
                 f.write(dockerfile_text)
+        
+        from ._cache import calculate_recursive_dir_hash, try_read_value_from_cache, write_value_to_cache
```
Comment (hongye-sun) [inline sdk/python/kfp/containers/_build_image_api.py]: nit: Could you move this import to the top of the file?

File: sdk/python/kfp/containers/_build_image_api.py [hunk 2]
```diff
@@ -111,11 +111,22 @@ def build_image_from_working_dir(image_name: str = None, working_dir: str = None
...
 
         if builder is None:
             builder = _get_default_image_builder()
-        return builder.build(
+        image_name = builder.build(
             local_dir=context_dir,
             target_image=image_name,
             timeout=timeout,
         )
+        if image_name:
```
Comment (hongye-sun) [inline sdk/python/kfp/containers/_build_image_api.py]: nit: just curious, in what circumstance that the image_name is empty?

Comment (Ark-kun) [reply sdk/python/kfp/containers/_build_image_api.py]: I should not be empty for `ContainerBuilder` now (previously it was returning `False` on errors). But technically there can exist a misbehaving builder that does not return anything.

File: sdk/python/kfp/containers/_build_image_api.py (hunk 1 above)
Comment (Ark-kun) [reply sdk/python/kfp/containers/_build_image_api.py]: I'll move it.

Comment (Ark-kun) [reply sdk/python/kfp/containers/_build_image_api.py]: Done.

**Pull request merged**